### ISSUE TITLE
make SocketType.type do the right thing by default

### ISSUE THREAD

**Issue by njsmith** [python-trio/trio#422]
Python 3.7 is fixing a long standing and annoying bug in the handling of socket object `type` attributes:

https://bugs.python.org/issue32331#msg308604

We have some delicate workarounds for this ourselves, but they're hidden because I wanted to see what the upstream API would look like. Now that we know what the fix will look like, we should copy it.

**Comment by njsmith** [comment]
Whoops, right, that's my fault for being unclear, sorry!

And right, we don't want to get rid of `real_socket_type` entirely, because we still need to work on 3.5 and 3.6. Instead, the idea would be to make Trio's socket object – on all versions – act like Python's socket object acts in 3.7+. And conveniently, Python's socket object in 3.7+ sets its `.type` attribute to be the same thing that Trio is already calculating with its `real_socket_type` hack.

So basically this means defining the `.type` attribute on *trio* socket objects to be something like `real_socket_type(self._sock.type)`.

And then we could go through all the places in trio that currently use `real_socket_type` and switch them to using `trio_socket.type` (it looks like there are currently 3 places, and they all have a trio socket object available).

Does that make sense?

**Comment by scottbelden** [comment]
Yes, that definitely makes sense. I'll see if I can put this together. Thanks!
